### PR TITLE
New syntax for travis notifications

### PR DESCRIPTION
**Pull request by shoogle**
The notification syntax appears to have changed slightly since before, but that wasn't the only problem. The notification section is a custom Travis command. It doesn't run in a bash shell so the environment variables are not set. If you want to obscure the email address and IRC channel then you need to use an [encrypted string](https://docs.travis-ci.com/user/encryption-keys/#Notifications-Example) instead.

### CODE REVIEW

File: .travis.yml [hunk 1]
```diff
@@ -112,18 +112,22 @@ matrix:
...
+      - "build@musescore.org"
     on_success: change
     on_failure: always
   irc:
     channels:
-      - $NOTIFICATION_IRC
+      - "chat.freenode.net#musescore"
     on_success: change
     on_failure: change
+    template: # https://docs.travis-ci.com/user/notifications/#IRC-notification
```
Comment (lasconic) [inline .travis.yml]: That's way too long on IRC. Any way to have it less verbose?

Comment (shoogle) [inline .travis.yml]: Too many lines, or the line itself is too long? If it's the latter then it's not actually that long once the variables get substituted in. E.g.:

<pre>
Build #174 for musescore/MuseScore failed after 2 min 46 sec. Author: shoogle
Branch: fix-travis-notifications, Commit: fb007e0 "New syntax for travis notifications"
Changes: https://github.com/shoogle/MuseScore/compare/09631ec750f1...fb007e0ac57d
Build log: https://travis-ci.org/shoogle/MuseScore/builds/142512037
</pre>

Travis' default notification template is:
<pre>
musescore/MuseScore#174 (fix-travis-notifications - fb007e0 - shoogle): The build failed.
Change view: https://github.com/shoogle/MuseScore/compare/09631ec750f1...fb007e0ac57d
Build details: https://travis-ci.org/shoogle/MuseScore/builds/142512037
</pre>

Comment (lasconic) [inline .travis.yml]: Sure, let's add skip_join: true
The message is one click away, so we don' really need it. It's a notification after all, shorter is better. I would happily keep the default message. It also has the advantage to be default and so similar in other projects, and so easier to parse visually.

Comment (shoogle) [inline .travis.yml]: Fair enough. I would have though the message would be the most important thing since everything else is just random digits, but it's your call.

**Comment by lasconic** [comment]
Thanks

**Comment by ericfont** [comment]
I protest, why are we now hardcoding the email address inside .travis.yml?  This causes my personal travis to spam IRC and build@musescore.org. 

In March committed 40426713b6e013b68965cffc259cb4bafe3d3206 allowed me to use environment variables to allow me to use my own travis and direct test failures to my own email.

What is the proper way that I'm supposed to use now?  Adding .travis.yml to my .gitignore means .gitignore wants to be tracked, and .git/info/excludes does not seem to be helping me (I guess because .travis.yml is already tracked).

**Comment by lasconic** [comment]
Without this PR, we were receiving no notification, none from your fork (which was OK) but also none from the master branch (which was bad...). 
Shoogle finds out that environment variables don't work in the notification section. Was it working for you? You received your own email? @shoogle any reference to back your claim? 

We can use encrypted string apparently. If we do that, it will stop email from your fork since you can't decode the string but you will not be able to get notification either...

**Comment by ericfont** [comment]
In the past, at least March-April yes I would get email notifications when my travis builds would fail, but I hadn't used it since then.

I'm fine not getting the email notification.

But I'm still spamming IRC.

**Comment by shoogle** [comment]
@lasconic can put the email and IRC channel in a [secure variable](https://docs.travis-ci.com/user/encryption-keys/#Notifications-Example) to prevent spam, as was done [here](https://github.com/musescore/MuseScore/blob/master/.travis.yml#L8-L9). This is the ideal solution from everyone's point of view, but must be done by @lasconic.

Evidence environment variables don't work for notifications:
- https://github.com/travis-ci/travis-ci/issues/2711
- https://github.com/travis-ci/travis-ci/issues/1094

Also, I tested it with my own email and a private IRC channel before submitting the PR.

@ericfont if you decide you do want notifications then do this each time you pull from upstream:
```bash
git update-index --assume-unchanged .travis.yml
```
If you add it to your [post-checkout script](https://github.com/musescore/MuseScore/tree/master/build/git/hooks) then that will be done for you automatically. You may also need to add `.travis.yml` to your .git/info/excludes file.

**Comment by ericfont** [comment]
Shoogle must be mistaken, since I'm clearly able to receive email notifications...here is my email inbox with travis notifications:

screenshot 43

Last one I received on before this PR was on July 5 and it seemed to work just fine as always, but then after this PR I stopped getting notifications.

Whoever wrote [this comment](https://github.com/travis-ci/travis-ci/issues/2711#issuecomment-161100623) must have been mistaken.

Shoogle, did you remember to add your email address that you use for email to your github settings of verified email addresses?

I was also doing some testing...and I may not be %100...but it seems the following works:

<pre>notifications:
  recipients:
    - $NOTIFICATION_EMAIL
  email:
    on_success: always
    on_failure: always</pre>

However if have the notification environment variable inside a recipient block inside the email block like the following, then the notifications do not work:

<pre> notifications:
   email:
     recipients:
       - $NOTIFICATION_EMAIL
    on_success: always
    on_failure: always
</pre>

(I don't know, but that might be the cause of why that commenter said it wasn't working??) 

If put email inside encrypted variable, then is it going to be slightly harder whenever for anyone setting up on a new computer who want to get notifications, but I guess I can manage.  It just seems like user-defined environment variables are the simple are correct way to handle this.

I'm curious, is build@musescore.org receiving notifications since this PR?

**Comment by shoogle** [comment]
@ericfont Yes, I used a verified email and I checked that it worked by trying it hard-coded. Very strange... the syntax you report working is the same as it was before my PR when @lasconic said it wasn't working.

**Comment by ericfont** [comment]
ok.  I'm wondering if there is some bug on travis's end or somewhere else in delivery that needs to be reported.

**Comment by lasconic** [comment]
Yes build@musescore.org does receive email currently.

**Comment by ericfont** [comment]
Well I can't come up with any logical explanation at this point except sometimes for some people, email via environment variable doesn't work.

shoolgle, one last dummy check, you are aware that the on_success: change and on_failure: change only report 

lasconic, was there any specific time when the build@musescore.org stopped working?

**Comment by ericfont** [comment]
I have a theory: Environment variables for emails actually don't work, but the emails were being sent to me because I was the author in the git commit.

https://docs.travis-ci.com/user/notifications/#How-is-the-build-email-receiver-determined%3F

> **By default, a build email is sent to the committer and the author**, but only if they have access to the repository the commit was pushed to. This prevents forks active on Travis CI from notifying the upstream repository’s owners when they’re pushing any upstream changes to their fork. It also prevents build notifications from going to folks not registered on Travis CI.

The email address is then determined based on the email address in the commit, but only if it matches one of the email addresses in our database. We synchronize all your email addresses from GitHub, solely for the purpose of build notifications.

> **The default can be overridden in the .travis.yml** as shown above. If there’s a setting specified, Travis CI only sends an emails to the addresses specified there, rather than to the committer and author.

**Comment by lasconic** [comment]
No mail on build@musescore.org since Oct 7 2015.

**Comment by lasconic** [comment]
So... let's sum it up. 
1. We can ask travis to expose environment variable for notification and wait... or ask them on #travis for a better solution.
2. In the meantime, we can use encrypted variables. I wonder if the committer will get an email on PR, since the variables will ne be decrypted...

**Comment by ericfont** [comment]
I don't think the commiter will get an email if the variable is encrypted.  I think the presense of the email notification line will override the default.  But won't know for sure until try it out...

Folks, I'm sorry for being incorrect on this issue...

**Comment by shoogle** [comment]
@ericfont no worries - it's not intuitive at all!

**Comment by ericfont** [comment]
:)

@lasconic "I wonder if the committer will get an email on PR, since the variables will ne be decrypted..."

That is actually correct!  I did a test:

<pre>
   email:
     recipients:
       - secure: "ThisIsNotActuallyReallyAnEncryptedEmail"
     on_success: always		      on_success: always
     on_failure: always		      on_failure: always
</pre>

Which sent me an email.  Same thing happened when I reversed 

<pre>
   recipients:
     - secure: "ThisIsNotActuallyReallyAnEncryptedEmail"
   email:
     on_success: always		      on_success: always
     on_failure: always		      on_failure: always
</pre>

So seems like encrypted address is way to go...will fall back to git commiters email if can't decrypt, which makes it convienint for people like me who want to get notified when i build on my own travis.

There is one thing left I'm going to test: that is whether recipients should be inside or outside or if doesn't matter...

**Comment by ericfont** [comment]
I've determined that putting recipients inside the email block is the correct way.  I determined this by using another of my email addresses as recipient, and tried once with that recipient block inside and tried once outside the email block.  When recipients was outside the email block, travis fell back to my commit's email address.  When recipients was inside the email block, travis used what was in recipients from the .travis.yml.

So in summary, the proper way to do this email is:

<pre>
   email:
     recipients:
       - secure: "LasconicGeneratesEncryptedStringAndPutsItHere"
</pre>

**Comment by lasconic** [comment]
Thank you both for the investigation. It's done on both active branches, master and 2.0.4. Let's see.
